common/src/margins.rs

```diff
@@ -21,6 +21,39 @@ pub fn fix_margins(
 
     let margin_coord = CoordsWithMargins::margin_coord(dimension, direction.sign);
     let edge_coord = CoordsWithMargins::edge_coord(dimension, direction.sign);
+
+    // If two solid chunks are both void or both non-void, do nothing.
+    if voxels.is_solid()
+        && neighbor_voxels.is_solid()
+        && (voxels.get(0) == Material::Void) == (neighbor_voxels.get(0) == Material::Void)
+    {
+        return;
+    }
+
+    // If the neighbor is solid and consistent with the edge of the current chunk, do nothing.
+    if !voxels.is_solid() && neighbor_voxels.is_solid() {
+        let neighbor_is_void = neighbor_voxels.get(0) == Material::Void;
+        if all_voxels_at_edge(dimension, voxels, direction, |m| {
+            (m == Material::Void) == neighbor_is_void
+        }) {
+            return;
+        }
+    }
+
+    // If the current chunk is solid and consistent with the edge of the neighbor, do nothing.
+    if voxels.is_solid() && !neighbor_voxels.is_solid() {
+        let is_void = voxels.get(0) == Material::Void;
+        if all_voxels_at_edge(
+            dimension,
+            neighbor_voxels,
+            neighbor_axis_permutation * direction,
+            |m| (m == Material::Void) == is_void,
+        ) {
+            return;
+        }
+    }
```

Comment:
nit: dedup? `for (a, b) in [(&*voxels, &*neighbor_voxels), (&*neighbor_voxels, &*voxels)]`

Reply:
Potentially, but for readability reasons, I'll probably avoid deduping here. The issue is that I also would need to distinguish between `direction` and `neighbor_axis_permutation * direction`.